### PR TITLE
black box tuning based on cma

### PR DESCRIPTION
**Pull request by zhiyongLiu1114**
### PR types

New features
### PR changes

Models 
### Description

A new method of prompt learning based on cma which is derivative-free, thus language model could be used as a service. Due to it is derivative-free, less gpu resources are required and less hard disks are needed while only prompt will be saved.

### CODE REVIEW

File: examples/few_shot/Black-Box-Tuning/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+
+
+[Black-Box Tuning for Language-Model-as-a-Service](https://arxiv.org/abs/2201.03514)
+
+
+## 算法简介
+
+这是一种黑盒调优框架，通过无导数优化对输入文本前面增加的连续提示语进行优化，当标记样本数量较少时可以取得较好效果。
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/README.md]: 建议增加更详细的模型介绍或者模型结构图，便于使用者快速了解。

File: examples/few_shot/Black-Box-Tuning/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+## 代码结构及说明
+```
+|—— train.py # 策略的训练、评估主脚本
+|—— data.py # 加载本地数据集脚本，以tnews为例
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/README.md]: 建议增加数据格式准备等相关的详细准确描述，提交目录中的数据集似乎不是 tnews。

File: examples/few_shot/Black-Box-Tuning/data.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        example["input_text"] = "%s . %s . It was %s ." % (prompt, example["text"], tokenizer.mask_token)
+        example["target_text"] = label2text[example["labels"]]
+    else:
+        example["input_text"] = "%s . It was %s ." % (example["text"], tokenizer.mask_token)
+        example["target_text"] = label2text[example["labels"]]
+
+    input_encodings = tokenizer.encode(example["input_text"], return_attention_mask=True, max_seq_len=max_seq_length)
+    target_encodings = tokenizer.encode(example["target_text"], add_special_tokens=False, max_seq_len=max_seq_length)
+    mask_pos = input_encodings["input_ids"].index(tokenizer.mask_token_id)
+    # encodings = {
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/data.py]: 请删除提交文件中非必要的注释。

File: examples/few_shot/Black-Box-Tuning/dev.tsv [hunk 4]
```diff
@@ -0,0 +1,32 @@
+calculating fiend 	positive
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/dev.tsv]: 建议增加该数据集的介绍。

File: examples/few_shot/Black-Box-Tuning/model.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+from paddle import nn
+
+from paddlenlp.transformers.ernie.modeling import (
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/model.py]: 这里 ERNIE 模型建议尽量使用框架 API，方便后续随框架一起升级维护。可以总结一下这里需要支持的功能需求，我们评估一下看是否能支持。

File: examples/few_shot/Black-Box-Tuning/model.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+class ErnieLMHead(nn.Layer):
+    """Roberta Head for masked language modeling."""
+
+    def __init__(self, config, embeddings):
+        super().__init__()
+        self.dense = nn.Linear(config["hidden_size"], config["hidden_size"])
+        self.layer_norm = nn.LayerNorm(config["hidden_size"])
+        self.activation = getattr(nn.functional, config["hidden_act"])
+        # self.decoder_weight = self.create_parameter(
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/model.py]: 建议删除非必要的注释。

File: examples/few_shot/Black-Box-Tuning/train.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+import paddle
+from data import convert_example, create_dataloader, load_data, transform_data
+from model import P_ErnieForMaskedLM
+from sklearn.metrics import f1_score
+from utils import file_path, label2text
+
+from paddlenlp.data import Pad, Stack, Tuple
+from paddlenlp.transformers.ernie.tokenizer import ErnieTokenizer
+
+parser = argparse.ArgumentParser()
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/train.py]: 请在`add_argument`时说明参数名的具体含义，并在 README 文档中增加相关介绍。

File: examples/few_shot/Black-Box-Tuning/train.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+random.seed(seed)
+np.random.seed(seed)
+paddle.seed(seed)
+
+
+class LMForwardAPI:
+    def __init__(self, model_name="ernie-3.0-medium-zh", n_prompt_tokens=50, loss_type="ce"):
+
+        self.model = P_ErnieForMaskedLM.from_pretrained("ernie-3.0-medium-zh", n_prompt_tokens=n_prompt_tokens)
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/train.py]: 这里传入 `model_name` 参数并未正常起作用，请确保所有配置参数能够按预期修改代码运行结果。

File: examples/few_shot/Black-Box-Tuning/train.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+        return loss, perf
+
+    def eval(self, prompt_embedding=None):
+
+        self.num_call += 1
+        if prompt_embedding is None:
+            prompt_embedding = self.best_prompt
+
+        bsz = len(train_data["input_ids"])  # for test data
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/train.py]: 建议使用完整句子描述代码和参数，避免使用含义模糊的描述，便于使用者理解。

File: examples/few_shot/Black-Box-Tuning/README.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+|—— model.py # BBT 模型
+|—— utils.py # 一些配置变量
+|—— run.sh # 一键训练及评估脚本
+
+```
+
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/README.md]: 建议在文档中增加代码执行效果相关的表格，列出具体数据集上的测试结果，便于后续正确性维护。

File: examples/few_shot/Black-Box-Tuning/train.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+parser.add_argument("--k_shot", default=16, type=int)
+parser.add_argument("--batch_size", default=16, type=int)
+parser.add_argument("--budget", default=20000, type=int)
+parser.add_argument("--popsize", default=20, type=int)
+parser.add_argument("--bound", default=0, type=int)
+parser.add_argument("--sigma", default=1, type=float)
+parser.add_argument("--print_every", default=50, type=int)
+parser.add_argument("--eval_every", default=100, type=int)
+parser.add_argument("--device", default="gpu:1", type=str)
+parser.add_argument("--alg", default="CMA", type=str)
```
Comment (LemonNoel) [inline examples/few_shot/Black-Box-Tuning/train.py]: 该参数似乎并未使用，请检查类似情况并删除或者修改。

File: examples/few_shot/Black-Box-Tuning/train.py (hunk 7 above)
Comment (LemonNoel) [reply examples/few_shot/Black-Box-Tuning/train.py]: 请增加以下三个参数，使 CI 测试能够正常运行，参数名称及对应功能如下：
- max_steps：最大执行步数
- save_steps：模型保存步数间隔
- output_dir：模型参数保存目录

**Review by LemonNoel — CHANGES_REQUESTED** [review]
CodeStyle 问题可通过以下方式解决，
1. 安装依赖
```
pip install pre-commit 
```
2. 在 PaddleNLP 目录下安装 pre-commit
```
cd PaddleNLP
pre-commit install
```
3. 在每个提交文件增加一个空格，重新 `git add` + `git commit` 从而触发代码风格检测逻辑，并按照反馈修改代码实现。
